### DISCUSSION TITLE
Consider referencing SUS instead of POSIX in future documentation.

### DISCUSSION THREAD

**Discussion by SirWumpus** [ioccc-src/mkiocccentry#849] [General]
While I am a great believer in POSIX.1 and .2 standards (intersection of BSD & SysV), XOpen came along to muddy the waters to ~~document~~ standardise the union of BSD and SysV APIs.  This later became the [Single Unix Specification](https://en.wikipedia.org/wiki/Single_UNIX_Specification), essentially POSIX plus Curses with lots of extra dangly bits.  

Q: Should future Contest Rules and Guidelines refer to SUS 2018 instead of POSIX?

One reason for this is simply the availability of the [Online Base Spec (& download)](https://pubs.opengroup.org/onlinepubs/9699919799.2018edition/) and [Curses Issue 7](https://pubs.opengroup.org/onlinepubs/9699909599/toc.pdf) that can be consulted by contestants (vs their OS man pages that might be divergent).  Kinda nice to play with the same round peg in square hole.

**Comment by lcn2** [comment]
Interning concept, @SirWumpus.

When we use the term POSIX in the guidelines, we use POSIX-like or we use P1003.1/P1003.2.  We do not specify a particular standard version or standard date.

The problem with using SUS 2018 is that few systems that are registered as being in compliance.  At best going back to [UNIX 03](https://unix.org/version3/online.html) would yield a few more registered UNIX systems.

Perhaps IOCCC should just refer the term [Single UNIX Specification-like](https://en.wikipedia.org/wiki/Single_UNIX_Specification) and point to the Wikipedia page and let people find their own standards documents if the feel the need to look up something in greater detail?

**Reply by SirWumpus** [reply]
Hmm.  I wasn't really consider a specific standard date (it just so happened to be the last SUS and also available online).  And the fact that systems don't get certified just tells you they don't want to pay the fee (I wouldn't), yet a common standard is still a great step towards perfection and enlightenment.  

SIDEBAR: reading the Wikipedia page, it seems `pax(1)`  is a common hold up; but it could be argued that it was created for a 1980's / 1990's problem and its need today is much less and could be marked "optional", given many of the enhancements to `tar(1)` and newer archive tools.  (They never did document `shar` and `unshar` :-) ).

> Perhaps IOCCC should just refer the term [Single UNIX Specification-like](https://en.wikipedia.org/wiki/Single_UNIX_Specification) and point to the Wikipedia page and let people find their own standards documents if the feel the need to look up something in greater detail?

How about leave all the existing references to POSIX as-is (given it is core SUS anyway) and provide the three links (Wikipedia, SUS Base Spec, Curses) as some helpful reference links as to acceptable reference material and POSIX superset?  Or something to that effect.  Explain how obfuscation is at the core of standards.

**Reply by lcn2** [reply]
> How about leave all the existing references to POSIX as-is (given it is core SUS anyway) and provide the three links (Wikipedia, SUS Base Spec, Curses) as some helpful reference links as to acceptable reference material and POSIX superset? Or something to that effect. Explain how obfuscation is at the core of standards.

Well @SirWumpus, we almost certainly do not want to or specific as to use the "P1003.1/P1003.2" notation.

Instead we can say "Unix-like" in the guidelines with a link to an FAQ of a question along the lines of "What is intended by the term Unix-like".  The answer to that FAQ can be text that mentions, SUS Base Spec, Curses, POSIX, Linux, FreeBSD, etc.  That FAQ answer can suggest some reference links (as you suggest) for someone to consider AND at same time suggest that being portable (as opposed to to dtrict adherence to some given standard or conforming to specific distribution guideline) is a **virtue** but **not** mandatory requirement.

We can stress that particular meaning of "Unix-line" is just a guideline.  And that guidelines are suggestions (most suggestions are good to consider, some of them are fun, some of them are traditional, a few of them are downright silly) and that guidelines are NOT rules.

... Well some sort of FAQ like that.  This was just a "off the top of our heads" test.  We will work on the FAQ language, add it to the FAQ and mention that in this discussion and invite corrections and improvements if needed.

**Reply by SirWumpus** [reply]
`This FAQ may have <scotish accent>great tracks of land</> on offer.`

**Reply by SirWumpus** [reply]
Humour aside. I can see that you don't want to necessarily limit contestant creativity. Yet do you need a separate FAQ? The Guidelines are already somewhat a Rationale and FAQ about the Rules. Now an FAQ about the Guidelines and the Rules?

**Reply by lcn2** [reply]
> Humour aside. I can see that you don't want to necessarily limit contestant creativity. Yet do you need a separate FAQ? The Guidelines are already somewhat a Rationale and FAQ about the Rules. Now an FAQ about the Guidelines and the Rules?

The [new FAQ](https://github.com/ioccc-src/temp-test-ioccc/blob/master/faq.md) is a set generic frequently asked questions relating to the overall contest.  They are not specific to a single contest.  They provide background information to the curious. 

Rules and guidelines are specific to a contest. They can stand on their own.  While the rules and guidelines can have external links to stuff, such external links are for illustrative and auxiliary purposes only.  The text that the rules and guidelines contain is what matters.

**Reply by lcn2** [reply]
We are still thinking of how to incorporate the pending rules, guidelines into the web site.

**Reply by SirWumpus** [reply]
Enjoy your `twinning` tea. 

Is there any point changing the Rules & Guidelines document format?  Maybe as a .pdf file that has a background watermark like `UNOFFICIAL` , `This is NOT a secret (yet).` , `Rule 6`, `This watermark was left by Chongo's Twinnings tea mug.` or similar Python-esque nonsense.

Also don't you normally keep a copy of the R&G with each year?  Depending how you layout the archive of past winners for public viewing, create a place holder year directory, eg. "Examples" or "Future" - with the next set (working copy) of unofficial R&G and other stuff subject to change.

**Comment by lcn2** [comment]
Just an a FYI / Update @SirWumpus: 

The new web site will have a [next directory](https://ioccc-src.github.io/temp-test-ioccc/next/index.html) with a [next/rules.html](https://ioccc-src.github.io/temp-test-ioccc/next/rules.html) and a [next/guidelines.html](https://ioccc-src.github.io/temp-test-ioccc/next/guidelines.html) pages.

Right now, they are just copies of the 2020 set.  However as soon as the web site design settles down, we plan to update those 2 pages to incorporate suggestions, such the general idea of this suggestion.

This will also make it easier for people to comment on and make suggested changes to the IOCCC Rules and Guidelines for the next ICOCC.

In regards to this Discussion specifically:

I general, we plan to more away from explicit _POSIX_ and instead consider the [Single UNIX Specification](https://en.wikipedia.org/wiki/Single_UNIX_Specification) such as perhaps [The Single UNIX Specification Version 4](https://unix.org/version4/overview.html).

**Comment by lcn2** [comment]
FYI @SirWumpus, "[FAQ 0.4: What platform should I assume for my entry?](https://ioccc-src.github.io/temp-test-ioccc/faq.html#SUS)" now refers to SUS on the [experimental IOCCC temp-test-ioccc](https://ioccc-src.github.io/temp-test-ioccc/) web site.

**Reply by SirWumpus** [reply]
Acceptable solution.

**Reply by lcn2** [reply]
> Acceptable solution.

There will be a mention in the guidelines as well.

**Comment by lcn2** [comment]
We believe we have addressed this discussion topic and will now close it with thanks to @SirWumpus for raising the awareness of this topic.
